### ISSUE TITLE
`clear(..)` supports a sync mode

### ISSUE THREAD

**Issue by sugarraysam** [bpftrace/bpftrace#3549]
### What reproduces the bug? Provide code if possible.

I have a map, which I want to delete immediately. The context is I use two maps `@curr` and `@prev`, and I am always writing to `@curr`, and at set intervals, I take the data in `@curr` and put it in `@prev`. For this to happen, I want to be able to `clear(@curr)`, but this creates issues as `clear(...)` only supports async deletion.

```bpftrace
#!/usr/bin/bpftrace

// write some data to current
usdt:/some/probe {
    @curr[arg1, arg2]++;
}

// at set interval we rotate buckets
interval:s:10 {
    // we are force to iterate through @curr to delete every entry manually
    // this is the only way to do sync deletion
    for (kv : @curr) {
        delete(@curr[$kv.0]);
    }
    
    <...>
}
```

### `bpftrace --info` output

System
  OS: Linux 5.19.0-0_fbk12_hardened_11583_g0bef9520ca2b #1 SMP Fri Feb 2 17:56:12 PST 2024
  Arch: x86_64

Build
  version: v0.21.0-80ca
  LLVM: 17.0.4
  bfd: yes
  liblldb (DWARF support): yes
  libsystemd (systemd notify support): no

Kernel helpers
  probe_read: yes
  probe_read_str: yes
  probe_read_user: yes
  probe_read_user_str: yes
  probe_read_kernel: yes
  probe_read_kernel_str: yes
  get_current_cgroup_id: yes
  send_signal: yes
  override_return: yes
  get_boot_ns: yes
  dpath: yes
  skboutput: yes
  get_tai_ns: yes
  get_func_ip: yes
  jiffies64: yes
  for_each_map_elem: yes

Kernel features
  Instruction limit: 1000000
  Loop support: yes
  btf: yes
  module btf: yes
  map batch: yes
  uprobe refcount (depends on Build:bcc bpf_attach_uprobe refcount): yes

Map types
  hash: yes
  percpu hash: yes
  array: yes
  percpu array: yes
  stack_trace: yes
  perf_event_array: yes
  ringbuf: yes

Probe types
  kprobe: yes
  tracepoint: yes
  perf_event: yes
  kfunc: yes
  kprobe_multi: yes
  uprobe_multi: no
  raw_tp_special: yes
  iter: yes

**Comment by sugarraysam** [comment]
Misclassified this my mistake, should be a "Feature request"

**Comment by danobi** [comment]
Similar to https://github.com/bpftrace/bpftrace/issues/3028 but this has less concerns about overload the ring buffers. Fairly straightforward and sensible to add an optional named parameter to `clear()`.

I asked @4ast and unfortunately it won't be much faster to try and add kernel support for this. There's always gonna be a loop somewhere. So for this implementation, we'd probably just want to loop over all elements and delete inline if the named param is true.

**Comment by amscanne** [comment]
There are tricks that can be used for structures like this to defer work but still present partially consistent views. It may be worthwhile considering having bpftrace manage these under the hood (which would be part of the value in using it!).

For example, consider having a global generation number for each map (and the value stored is `(gen, actual value)`). You always read the current generation, and on lookups, something with `gen <= current_gen` is treated as not present. All inserts and updates will set the generation to the current one as well. The `clear` operation is then merely `current_gen++`. Since `clear` should be the only `O(N)` operation, all individual ops become marginally more expensive (since you need to check against the generation), but all map ops are then `O(1)`.

You could add a second counter to make prints partially consistent as well against deletes.

**Comment by 4ast** [comment]
> For example, consider having a global generation number for each map (and the value stored is `(gen, actual value)`).

if clear() has to be that fast and happens often then we can do this trick in the kernel too, but there is a bit of a challenge with map->max_entries that was introduced to limit the amount of memory used by maps and size the hash map bucket array at init time. With generation count we'd have to reuse older elements on inserts which may get tricky and expensive.
Worth exploring, I guess.

**Comment by jordalgo** [comment]
After some offline discussion, we want to just make clear synchronous and not support a sync "mode". This could be done fairly easily in the map_sugar pass where we turn a call to `clear` into a loop over the map that calls delete.

Example:
```
clear(@a);
```
Would become
```
for ($kv : @a) { delete(@a, $kv.0); } }
```

**Comment by lechedesnatada23** [comment]
Hi! Following up on the direction @jordalgo set out (make `clear()`
synchronous, no sync "mode") — I've been working on an implementation and
want to check I'm not stepping on anyone's toes before opening anything.

I saw @Aevil1's draft #4319 (June 2025, inactive since), which rewrites
`clear(@map)` into `for ($kv : @map) { delete(@map[$kv.0]); }` in
`map_sugar`. From what I can tell, that approach hits a limitation: the type
checker rejects `for` loops over maps with per-cpu aggregate values
(`IsMapIterableTy()`), so `clear()` would fail to compile for
`hist`/`lhist`/`stats`/`tseries` maps — and there are runtime tests
explicitly covering those ("hist/lhist/stats/avg/tseries can be cleared" in
`tests/runtime/other`). Happy to be corrected if I missed part of the draft.

So I implemented it the same way `delete()` works today: `clear` is a stdlib
macro backed by a small C helper (`__clear()` in `stdlib/map/map.bpf.c`)
that walks the map with `bpf_for_each_map_elem()` and deletes each key
in-kernel — same pattern as the existing `__elem_count()`. That keeps
`clear()` working on every map type it supports today, and it is fully
synchronous, so `clear` events can no longer be dropped when the ring buffer
fills up (the #3924 concern). One inherent behavior change: since deletion
happens in-kernel, a `print()` emitted before `clear()` in the same block
sees an empty map.

It's complete locally with tests and docs (full unit suite green; runtime
tests pass except a few pre-existing environment failures that reproduce
identically on unpatched master).

How would you like to proceed?
1. I open my implementation as a draft PR for comparison (with credit to
   #4319 — happy to adopt any part of it),
2. I help finish/test #4319 instead, or
3. anything else you'd prefer.

I won't open a PR until I hear from you. Thanks!

(For transparency: I prepared this with the help of an AI coding agent
under my direction; the implementation and tests were built and run locally
on my machine.)

**Comment by viktormalik** [comment]
Hi @lechedesnatada23, thanks for picking this up! I think a new PR is the best for visibility, with a link to the original one for comparison. Once your PR is approved and merged, we'll just close the old one.

**Comment by lechedesnatada23** [comment]
Thanks @viktormalik! Opened #5323, with a link to #4319 for comparison as you suggested.
